src/Microsoft.AspNet.Mvc.Core/Formatters/MediaType.cs

```diff
@@ -0,0 +1,550 @@
+            // We check if the parsed media type has value at this stage when we have iterated
+            // over all the parameters and we know if the parsing was sucessful.
+            if (!parser.ParsingFailed)
+            {
+                return new MediaTypeSegmentWithQuality(
+                    new StringSegment(mediaType, start, parser.CurrentOffset - start),
+                    quality);
+            }
+            else
+            {
+                return default(MediaTypeSegmentWithQuality);
+            }
+        }
+
+        private static Encoding GetEncodingFromCharset(StringSegment charset)
+        {
+            if (charset.Equals("utf-8", StringComparison.OrdinalIgnoreCase))
+            {
+                // This is an optimization for utf-8 that prevents the Substring caused by
+                // charset.Value
+                return Encoding.UTF8;
+            }
+
+            try
+            {
+                // charset.Value might be an invalid encoding name as in charset=invalid.
+                // For that reason, we catch the exception thrown by Encoding.GetEncoding
+                // and return null instead.
+                return charset.HasValue ? Encoding.GetEncoding(charset.Value) : null;
+            }
+            catch (Exception)
+            {
+                return null;
+            }
+        }
+
+        private static string CreateMediaTypeWithEncoding(StringSegment mediaType, Encoding encoding)
+        {
+            return $"{mediaType.Value}; charset={encoding.WebName}";
+        }
+
+        private bool ContainsAllParameters(MediaTypeParameterParser setParameters)
+        {
+            var parameterFound = true;
+            MediaTypeParameter setParameter;
+            while (setParameters.ParseNextParameter(out setParameter) && parameterFound)
+            {
+                if (setParameter.HasName("q"))
+                {
+                    // "q" and later parameters are not involved in media type matching. Quoting the RFC: The first
+                    // "q" parameter (if any) separates the media-range parameter(s) from the accept-params.
+                    break;
+                }
+
+                // Copy the parser as we need to iterate multiple times over it.
+                // We can do this because it's a struct
+                var subSetParameters = _parameterParser;
+                parameterFound = false;
+                MediaTypeParameter subSetParameter;
+                while (subSetParameters.ParseNextParameter(out subSetParameter) && !parameterFound)
+                {
+                    parameterFound = subSetParameter.Equals(setParameter);
+                }
+            }
+
+            return parameterFound;
+        }
+
+        private struct MediaTypeParameterParser
+        {
+            private string _mediaType;
+            private int? _length;
+
+            public MediaTypeParameterParser(string mediaType, int offset, int? length)
```

Comment:
Should we call `mediaType` `buffer` or something similar? It's really not required to be a single media type, it's a string that contains a media type.